packages/core/src/browser/tree/tree-widget.tsx

```diff
@@ -153,7 +153,7 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
                     depth
                 }] as [string, TreeWidget.NodeRow];
             });
-            this.rows = new Map(rows);
+            this.rows = new Map(rows.filter(r => TreeNode.isVisible(r[1].node)));
```

Comment:
If the node is invisible, don't even load it into `this.rows` or the `map`. We don't want React to render an empty row for an invisible node.

Reply:
It looks expensive. It will iterate over all rows again. Also, it should be covered already by https://github.com/theia-ide/theia/blob/a09db28d039b993472a623b5f0dcdf29296e3f04/packages/core/src/browser/tree/tree-widget.tsx#L252 and https://github.com/theia-ide/theia/blob/a09db28d039b993472a623b5f0dcdf29296e3f04/packages/core/src/browser/tree/tree-widget.tsx#L465.
How can I reproduce it?

Reply:
@akosyakov you are right, I should have come up with a better solution. Let's me try changing other places to get the same end effect. 

To reproduce it, you would need to 1) pull this branch, 2) revert this line of code, 3) compile, 4) set preference 'workspace.supportMultiRootWorkspace' to true, 5) start theia, 6) open a folder, and 7) add one more folder to the workspace to make it 'multi root'.

What we *should* have seen at this point, is something like the below screenshot (more than one root folders in the navigator):
![screenshot from 2018-07-11 08-40-00](https://user-images.githubusercontent.com/37082801/42571806-0ae57c5c-84e6-11e8-87d5-2863c5477963.png)

However, we would see is (the black empty area above the `helloworld`, the first root folder)

![screenshot from 2018-07-11 08-44-04](https://user-images.githubusercontent.com/37082801/42571979-95c090be-84e6-11e8-8c92-8d242aa9efae.png)

and sometimes this (one node is rendered on top of another):

![screenshot from 2018-07-11 08-40-00](https://user-images.githubusercontent.com/37082801/42572044-c5f59c02-84e6-11e8-88a5-588b6d59fd0f.png)
 
What causes the problem is that, even if `renderNode()` (line 465) returns `null`, the `renderTreeRow()` (line 767) still inserts a DOM element, which takes space, and offsets its siblings.

Reply:
We could try to incorporate filtering of invisible nodes it in the iterator, so they gets filtered in the same pass when rows are constructed.

Reply:
In case the iterator shouldn't care, would this work ?
```ts
        if (root) {
            const depths = new Map<CompositeTreeNode | undefined, number>();
            const rows = [];
            let index = 0;

            for (const node of new TopDownTreeIterator(root, {
                pruneCollapsed: true,
                pruneSiblings: true
            })) {
                if (node.visible) {
                    const parentDepth = depths.get(node.parent);
                    const depth = parentDepth === undefined ? 0 : TreeNode.isVisible(node.parent) ? parentDepth + 1 : parentDepth;
                    if (CompositeTreeNode.is(node)) {
                        depths.set(node, depth);
                    }
                    rows.push(
                        [node.id, {
                            index: index++,
                            node,
                            depth
                        }] as [string, TreeWidget.NodeRow]
                    );
                }
            }
            this.rows = new Map(rows);
```
The for loop would replace the `Array.from` but still consume the iterator in one pass.

Reply:
@macechal-p sounds good if it works without any side-effects, please check the outline and problem trees as well

Reply:
@akosyakov done in the way that @marechal-p suggested. 

I extended that function in `FileNavigatorWidget` class, so the change only applies to the tree in navigator. outline and problem trees are not impacted at all.

Reply:
but now we have a lot of duplicate code? why don't change it in the tree-widget, we should not render invisible nodes in any tree

Reply:
sure. the changes are now only made to `doUpdateRows()` in `tree-widget`. Tested. outline and problem trees are still working with the change :)